workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/ScreenComposableFactory.kt

```diff
@@ -0,0 +1,233 @@
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.platform.ComposeView
+import androidx.compose.ui.platform.LocalLifecycleOwner
+import androidx.compose.ui.viewinterop.AndroidView
+import androidx.lifecycle.setViewTreeLifecycleOwner
+import com.squareup.workflow1.ui.Screen
+import com.squareup.workflow1.ui.ScreenViewFactory
+import com.squareup.workflow1.ui.ScreenViewHolder
+import com.squareup.workflow1.ui.ViewEnvironment
+import com.squareup.workflow1.ui.ViewRegistry
+import com.squareup.workflow1.ui.ViewRegistry.Key
+import com.squareup.workflow1.ui.WorkflowUiExperimentalApi
+import com.squareup.workflow1.ui.androidx.OnBackPressedDispatcherOwnerKey
+import com.squareup.workflow1.ui.show
+import com.squareup.workflow1.ui.startShowing
+import kotlin.reflect.KClass
+
+@WorkflowUiExperimentalApi
+public inline fun <reified ScreenT : Screen> ScreenComposableFactory(
+  noinline content: @Composable (
+    rendering: ScreenT,
+    environment: ViewEnvironment
+  ) -> Unit
+): ScreenComposableFactory<ScreenT> = ScreenComposableFactory(ScreenT::class, content)
+
+@PublishedApi
+@WorkflowUiExperimentalApi
+internal fun <ScreenT : Screen> ScreenComposableFactory(
+  type: KClass<in ScreenT>,
+  content: @Composable (
+    rendering: ScreenT,
+    environment: ViewEnvironment
+  ) -> Unit
+): ScreenComposableFactory<ScreenT> = object : ScreenComposableFactory<ScreenT> {
+  override val type: KClass<in ScreenT> = type
+
+  @Composable override fun Content(
+    rendering: ScreenT,
+    environment: ViewEnvironment
+  ) {
+    content(rendering, environment)
+  }
+}
+
+/**
+ * A [ViewRegistry.Entry] that uses a [Composable] function to display [ScreenT].
+ * This is the Compose-based analogue of
+ * [ScreenViewFactory][com.squareup.workflow1.ui.ScreenViewFactory]. But unlike
+ * [ScreenViewFactory], it is unusual for app developers to interact with this
+ * class directly. Most of the time you can subclass [ComposeScreen] and rely on
+ * the standard hidden [ScreenComposableFactory] that backs it.
+ *
+ * - See [ComposeScreen] for a more complete description of using Compose to
+ *   build a Workflow-based UI.
+ *
+ * - See [WorkflowRendering] to display a nested [Screen] from [ComposeScreen.Content]
+ *   or from [ScreenComposableFactory.Content]
+ *
+ * Use [ScreenComposableFactory] directly if you need to prevent your
+ * [Screen] rendering classes from depending on Compose at compile time.
+ *
+ * Example:
+ *
+ *     val fooComposableFactory = ScreenComposableFactory<FooScreen> { screen, _ ->
+ *       Text(screen.message)
+ *     }
+ *
+ *     val viewRegistry = ViewRegistry(fooComposableFactory, …)
+ *     val viewEnvironment = ViewEnvironment.EMPTY + viewRegistry
+ *
+ *     renderWorkflowIn(
+ *       workflow = MyWorkflow.mapRendering { it.withEnvironment(viewEnvironment) }
+ *     )
+ *
+ */
+@WorkflowUiExperimentalApi
+public interface ScreenComposableFactory<in ScreenT : Screen> : ViewRegistry.Entry<ScreenT> {
```

Comment:
outside of scope of this, but out of curiosity if we were wanting to add a compose overlay factory.

would that entail creating a new `ComposeOverlayDialogFactory` which implements `ScreenComposableFactory` & then adding to the [checks in ScreenComposableFactoryFinder](https://github.com/square/workflow-kotlin/pull/1146/files#diff-488dc3940ee9cf848df3a29160328f6b811dcca43c0f130d52d10d66fad40cfdR30) searching for `ComposeOverlayDialogFactory`?

or would that entail creating a new factory like this one, with it's relative factory finder?

Reply:
That would require:

- A Compose binding (`ScreenComposableFactory` implementation) for `BodyAndOverlaysScreen` -- an alternative to the existing `BodyAndOverlaysContainer`

- A new `OverlayComposableFactory` / `OverlayComposableFactoryFinder` pairing

That's assuming you weren't going to use any non-Compose dialog code at all.

If you were going to try to make it coexist with the `DialogComposableFactory` you would also need to enhance the logic in `withComposeInteropSupport()`. But that would only be feasible if your Compose version was still going to create `Dialog` instances, so that Compose-managed ones could interleave with legacy instances. Even then, you would somehow have to make `DialogCollator` et al agnostic between `Dialog` intances created via `Compose` v. those created the other way. No idea if that's practical.

Reply:
yeah exactly not going to use any non-compose dialog code at all - thanks for explaining!